### PR TITLE
[Jormun] adapt properly durations, distances and sections for ridesharing journeys

### PR DESCRIPTION
**Pull request by benoit-bst**
There was several problems with ridesharing journeys. 
_Durations_, _distances_, _sections type_ was not exact and we didn't use _street network information_ when it exist.

List of the fixes:

- **Blablalines** : the departures date time is the request datetime (because we haven't the information)
- **Instant system**: add a crowfly section with manhattan distance
- **All**: if we have a shape for _first section or last section_ we create street network section with all informations. Otherwise, it is a crowfly section
- **All**: if we have a shape with 2 coords, this is not a street network but a crowfly (Karos case)
- **All**: for all connector we fill _departure_date_time_ and _arrival_date_time_
- **All**: the sequence of _date time_ in the ridesharing journeys sections is consistent
- **All**: durations and distances are correct

Sorry for the reviewer. It is not easy to understand all stuff, but I assure you I tested in real conditions :cat: and it works

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/ridesharing/blablalines.py [hunk 1]
```diff
@@ -114,14 +115,39 @@ def _make_response(self, raw_json):
             res.distance = offer.get('distance')
             res.duration = offer.get('duration')
             res.ridesharing_ad = offer.get('web_url')
-            res.origin_pickup_duration = offer.get('departure_to_pickup_walking_time')
-            res.dropoff_dest_duration = offer.get('dropoff_to_arrival_walking_time')
 
-            res.pickup_place = rsj.Place(
-                addr='', lat=offer.get('pickup_latitude'), lon=offer.get('pickup_longitude')
+            # departure coord
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/blablalines.py:119]:
```suggestion
            # departure coord is absent in the offer and hence to be generated from the request  
```

File: source/jormungandr/jormungandr/scenarios/ridesharing/blablalines.py [hunk 2]
```diff
@@ -114,14 +115,39 @@ def _make_response(self, raw_json):
...
-                addr='', lat=offer.get('dropoff_latitude'), lon=offer.get('dropoff_longitude')
+
+            # drop off coord
+            dropoff_lat = offer.get('dropoff_latitude')
+            dropoff_lon = offer.get('dropoff_longitude')
+            dropoff_coord = Coords(lat=dropoff_lat, lon=dropoff_lon)
+
+            res.dropoff_place = rsj.Place(addr='', lat=dropoff_lat, lon=dropoff_lon)
+
+            # arrival coord
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/blablalines.py:143]: Modify comment same as above..

File: source/jormungandr/jormungandr/scenarios/ridesharing/karos.py [hunk 3]
```diff
@@ -111,7 +111,21 @@ def _make_response(self, raw_json):
...
             res.origin_pickup_duration = offer.get('departureToPickupWalkingTime')
+            res.origin_pickup_distance = int(offer.get('departureToPickupWalkingDistance'))
+            res.origin_pickup_shape = []
+            origin_pickup_shape = decode_polyline(offer.get('departureToPickupWalkingPolyline'), precision=5)
+            if origin_pickup_shape:
+                res.origin_pickup_shape.extend(
+                    (type_pb2.GeographicalCoord(lon=c[0], lat=c[1]) for c in origin_pickup_shape)
+                )
             res.dropoff_dest_duration = offer.get('dropoffToArrivalWalkingTime')
+            res.dropoff_dest_distance = int(offer.get('dropoffToArrivalWalkingDistance'))
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/karos.py]: As we can not guarantee the type and format of attributes values in any RS connector, isn't it better to transform in ridesharing_service_manager ? 
For example offer.get('dropoffToArrivalWalkingDistance') is float in karos should be transformed to int() as you did above.
I think we should transform value of mostly every attribute of ridesharing_journey (ridesharing_journey.py)

File: source/jormungandr/jormungandr/scenarios/ridesharing/karos.py [hunk 4]
```diff
@@ -141,8 +155,9 @@ def _make_response(self, raw_json):
             res.total_seats = None
 
             res.pickup_date_time = offer.get('driverDepartureDate')
-            if res.pickup_date_time is not None:
-                res.dropoff_date_time = res.pickup_date_time + offer.get('duration')
+            res.departure_date_time = res.pickup_date_time - res.origin_pickup_duration
+            res.dropoff_date_time = res.pickup_date_time + offer.get('duration')
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/karos.py:139]: Can we guarantee that res.pickup_date_time exists?

File: source/jormungandr/jormungandr/scenarios/ridesharing/karos.py [hunk 5]
```diff
@@ -200,14 +215,14 @@ def _request_journeys(self, from_coord, to_coord, period_extremity, limit=None):
         if resp:
             r = self._make_response(resp.json())
             self.record_additional_info('Received ridesharing offers', nb_ridesharing_offers=len(r))
-            logging.getLogger('stat.ridesharing.instant-system').info(
+            logging.getLogger('stat.ridesharing.karos').info(
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/karos.py:198]: :smile:

File: source/jormungandr/jormungandr/scenarios/ridesharing/ridesharing_service_manager.py [hunk 6]
```diff
@@ -275,29 +275,47 @@ def build_ridesharing_journeys(self, from_pt_obj, to_pt_obj, period_extremity):
...
             # start teleport section
             start_teleport_section = pb_rsj.sections.add()
             start_teleport_section.id = "section_{}".format(six.text_type(generate_id()))
-            start_teleport_section.type = response_pb2.CROW_FLY
             start_teleport_section.street_network.mode = response_pb2.Walking
             start_teleport_section.origin.CopyFrom(from_pt_obj)
             start_teleport_section.destination.CopyFrom(pb_rsj_pickup)
-            start_teleport_section.length = int(crowfly_distance_between(from_coord, pickup_coord))
+            if rsj.origin_pickup_shape:
+                if len(rsj.origin_pickup_shape) == 2:
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/ridesharing_service_manager.py:299]: Add a comment "If stree_network shape in the offer contains with only two points use default CROW_FLY section"

File: source/jormungandr/jormungandr/scenarios/ridesharing/ridesharing_service_manager.py [hunk 7]
```diff
@@ -343,36 +361,55 @@ def build_ridesharing_journeys(self, from_pt_obj, to_pt_obj, period_extremity):
...
 
             # end teleport section
             end_teleport_section = pb_rsj.sections.add()
             end_teleport_section.id = "section_{}".format(six.text_type(generate_id()))
-            end_teleport_section.type = response_pb2.CROW_FLY
             end_teleport_section.street_network.mode = response_pb2.Walking
             end_teleport_section.origin.CopyFrom(pb_rsj_dropoff)
             end_teleport_section.destination.CopyFrom(to_pt_obj)
-            end_teleport_section.length = int(crowfly_distance_between(dropoff_coord, to_coord))
+            if rsj.dropoff_dest_shape:
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/scenarios/ridesharing/ridesharing_service_manager.py:390]: Add a comment as above

File: source/jormungandr/jormungandr/street_network/utils.py [hunk 8]
```diff
@@ -47,6 +47,10 @@ def crowfly_distance_between(start_coord, end_coord):
     return EARTH_RADIUS_IN_METERS * 2.0 * math.asin(math.sqrt(lat_h + tmp * lon_h))
 
 
+def get_manhattan_duration(distance, speed):
```
Comment (kadhikari) [inline source/jormungandr/jormungandr/street_network/utils.py:50]: There are other get_manhattan_duration in jormungandr. Isn't it better to add one in /jormungandr/utils.py and delete other functions and their uses ?

File: source/jormungandr/tests/blablalines_routing_tests.py [hunk 9]
```diff
@@ -171,8 +178,10 @@ def test_basic_ride_sharing(self):
 
         assert rsj_sections[2].get('type') == 'crow_fly'
         assert rsj_sections[2].get('mode') == 'walking'
-        # For end teleport section we take dropoff to destination duration
         assert rsj_sections[2].get('duration') == 76
+        assert rsj_sections[2].get('departure_date_time') == '20120614T081935'
+        assert rsj_sections[2].get('arrival_date_time') == '20120614T082051'
+        # For end teleport section we take dropoff to destination duration
```
Comment (kadhikari) [inline source/jormungandr/tests/blablalines_routing_tests.py]: Why you moved this comment line here?

File: source/jormungandr/jormungandr/scenarios/ridesharing/karos.py (hunk 3 above)
Comment (benoit-bst) [reply source/jormungandr/jormungandr/scenarios/ridesharing/karos.py]: :+1: 
it is a good idea to cast with the good format within ridesharing_service_manager

File: source/jormungandr/jormungandr/scenarios/ridesharing/karos.py (hunk 4 above)
Comment (benoit-bst) [reply source/jormungandr/jormungandr/scenarios/ridesharing/karos.py:139]: yes, pickup_date_time is mandatory in Karos API

File: source/jormungandr/jormungandr/street_network/utils.py (hunk 8 above)
Comment (benoit-bst) [reply source/jormungandr/jormungandr/street_network/utils.py:50]: It is a good idea.
There is 3 pathes of manhattan function:
- jormun/utils
- scenario/helper_classes/fallback_duration
- jormun/street_network/utils

Maybe in an other PR i don't want to change everywhere in this PR

File: source/jormungandr/tests/blablalines_routing_tests.py (hunk 9 above)
Comment (benoit-bst) [reply source/jormungandr/tests/blablalines_routing_tests.py]: it is an oblivion

File: source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py [hunk 10]
```diff
@@ -141,22 +142,52 @@ def _make_response(self, raw_json):
...
-                res.pickup_place = rsj.Place(
-                    addr=from_data.get('name'), lat=from_data.get('lat'), lon=from_data.get('lon')
+                res.pickup_place = rsj.Place(addr=from_data.get('name'), lat=pickup_lat, lon=pickup_lon)
+
+                res.origin_pickup_distance = int(
+                    jormungandr.street_network.utils.crowfly_distance_between(departure_coord, pickup_coord)
+                )
+                # we choose to calculate with speed=1.12 the average speed for a walker
+                res.origin_pickup_duration = jormungandr.street_network.utils.get_manhattan_duration(
+                    res.origin_pickup_distance, 1.12
```
Comment (azime) [inline source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py]:
```suggestion
                    res.origin_pickup_distance, instance.walking_speed
```
No ?

File: source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py [hunk 11]
```diff
@@ -141,22 +142,52 @@ def _make_response(self, raw_json):
...
+                # arrival coord
+                lat, lon = to_coord.split(',')
+                arrival_coord = Coords(lat=lat, lon=lon)
+
+                res.dropoff_dest_distance = int(
+                    jormungandr.street_network.utils.crowfly_distance_between(dropoff_coord, arrival_coord)
+                )
+                # we choose to calculate with speed=1.12 the average speed for a walker
+                res.dropoff_dest_duration = jormungandr.street_network.utils.get_manhattan_duration(
+                    res.dropoff_dest_distance, 1.12
```
Comment (azime) [inline source/jormungandr/jormungandr/scenarios/ridesharing/instant_system.py]:
```suggestion
                    res.dropoff_dest_distance, instance.walking_speed
```

File: source/jormungandr/jormungandr/scenarios/ridesharing/klaxit.py [hunk 12]
```diff
@@ -111,7 +111,21 @@ def _make_response(self, raw_json):
...
+                    (type_pb2.GeographicalCoord(lon=c[0], lat=c[1]) for c in origin_pickup_shape)
+                )
             res.dropoff_dest_duration = offer.get('dropoffToArrivalWalkingTime')
+            res.dropoff_dest_distance = offer.get('dropoffToArrivalWalkingDistance')
+            res.dropoff_dest_shape = []
+            dropoff_dest_shape = decode_polyline(offer.get('dropoffToArrivalWalkingPolyline'), precision=5)
+            if dropoff_dest_shape:
+                res.dropoff_dest_shape.extend(
+                    (type_pb2.GeographicalCoord(lon=c[0], lat=c[1]) for c in dropoff_dest_shape)
+                )
```
Comment (azime) [inline source/jormungandr/jormungandr/scenarios/ridesharing/klaxit.py]: We can do refacto: the same code [her](https://github.com/CanalTP/navitia/pull/3316/files#diff-cd3659bd32fe17d791cc0d4026b844d9517aea30e16baeeb5f1385f6a6076939R113-R128)

**Review by kadhikari — APPROVED** [review]
:+1:  Great work.

**Review by azime — APPROVED** [review]

**Comment by azime** [comment]
:+1:

**Pull request merged**